src/main/scala/com/example/solution/services/Reservations.scala

```diff
@@ -0,0 +1,134 @@
+
+  import ReservationQueries._
+
+  private def areSeatsValid(picked: List[PickedSeat],
+                            available: List[AvailableSeat],
+                            seatsPerRow: Int): Boolean = {
+    val isSeatAlreadyTaken = picked.exists(ps => !available.contains(AvailableSeat(ps.row, ps.seatInRow)))
+    val leftSeats = available diff picked.map(ps => AvailableSeat(ps.row, ps.seatInRow))
+    val leftOneEmptySeat = leftSeats.groupBy(_.row).exists(tuple => {
+      val list = tuple._2
+      list.exists(seat => {
+        val seatInRow = seat.seatInRow
+        (seatInRow > 1 && seatInRow < seatsPerRow) &&
+          (!list.exists(as => as.seatInRow == seatInRow - 1) && !list.exists(as => as.seatInRow == seatInRow + 1))
+      })
+    })
+
+    !(isSeatAlreadyTaken || leftOneEmptySeat)
+  }
+
+
+  private def getSeats(seats: List[PickedSeat]): OptionT[F, NonEmptyList[PickedSeat]] =
+    OptionT(Sync[F].delay(seats.headOption match {
+      case Some(value) => Option(NonEmptyList(value, seats.tail))
+      case None => Option.empty[NonEmptyList[PickedSeat]]
+    }))
+
+
+  override def makeReservation(data: ReservationData): F[Option[MadeReservation]] = {
+
+    val userWithSeats = (for {
+      user <- OptionT(Sync[F].delay(User.mkUser(UserName(data.name), UserSurname(data.surname))))
+      rd <- OptionT(screenings.pickScreening(data.screeningId))
+      if rd.screeningTime.minus(15L, ChronoUnit.MINUTES).compareTo(LocalDateTime.now()) > 0
+
+      pickedSeats <- getSeats(data.pickedSeats)
+      seatsWithTickets <- OptionT(Sync[F].delay(pickedSeats.map(ps => for {
+        seat <- Option(ps)
+        ticket <- Ticket.fromString(ps.ticketType)
+      } yield (seat, ticket)).sequence)) if areSeatsValid(seatsWithTickets.map(_._1).toList, rd.availableSeats, rd.room.seatsPerRow)
+
+    } yield (user, seatsWithTickets)).value
+
+    userWithSeats.flatMap {
+      case None => Sync[F].delay(Option.empty[MadeReservation])
+      case Some(tuple) =>
+        val (user, seatsWithTickets) = (tuple._1, tuple._2.toList)
+        sessionPool.use { session =>
+          session.prepare(createReservation).use { cmd =>
+            val totalCost = seatsWithTickets.map(_._2.price).sum
+            cmd.execute((user, data.screeningId), totalCost).void
+          } *> session.execute(selectMaxId).map(_.head).flatMap(id => {
+            seatsWithTickets.traverse(tuple => {
+              val (ps, ticket) = tuple
+              session.prepare(createTakenSeat).use { cmd =>
+                cmd.execute((id, ps), ticket).void
+              }
+            })
+          }).map(_ => MadeReservation(
+            seatsWithTickets.map(_._2.price).sum,
+            LocalDateTime.now().plus(1L, ChronoUnit.HOURS))
+          )
+          //expiration date is obviously incorrect here
+        }.map(mr => Option(mr))
+    }
+  }
+}
+
+private object ReservationQueries {
+
+
+  val selectMaxId: Query[Void, Int] =
```

Comment:
let's say requirements are changed and now we have to use mongodb as a data store. How would you do that? What would you change to make the transition simpler?

Reply:
Since Skunk talks only to Postgres we would need to change a big chunk of code, I guess we could introduce a middle layer, an interface providing functions like selectMaxId() : Int etc, but makeReservation function is still very Skunk specific. I think the right way to approach it would be to either have another class implementing the mongodb version, or write
a 'high-level' function like 
`for {
_ <- createReservation(user,screeningId)
id <- selectMaxId()
_ <- seatsWithTickets.traverse((_,_) => createTakenSeat())
} yield ()
`(not sure if the code block is readable) that would not commit to any concrete implementation, and have functions like createReservation being provided by a service passed as an argument.